kv/kvclient/kvcoord: TestMuxRangeFeedDoesNotStallOnError failed

kv/kvclient/kvcoord.TestMuxRangeFeedDoesNotStallOnError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17942979?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17942979?buildTab=artifacts#/) on master @ [fa7b97990f45bdf3d7f7896b66379489934091c8](https://github.com/cockroachdb/cockroach/commits/fa7b97990f45bdf3d7f7896b66379489934091c8):


```
=== RUN   TestMuxRangeFeedDoesNotStallOnError
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/a38679e165d5c98145544b9bfab9e765/logTestMuxRangeFeedDoesNotStallOnError1337178714
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    dist_sender_rangefeed_test.go:389: condition failed to evaluate within 45s: from dist_sender_rangefeed_test.go:395: at least one non-local range required for test
    panic.go:626: -- test log scope end --
--- FAIL: TestMuxRangeFeedDoesNotStallOnError (49.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMuxRangeFeedDoesNotStallOnError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44970

kv/kvclient/kvcoord.TestMuxRangeFeedDoesNotStallOnError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17958424?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17958424?buildTab=artifacts#/) on master @ [0f2422bac446d1d297cec730ec7817aef46bd2be](https://github.com/cockroachdb/cockroach/commits/0f2422bac446d1d297cec730ec7817aef46bd2be):


```
=== RUN   TestMuxRangeFeedDoesNotStallOnError
I241128 18:43:29.346669 1 (gostd) rand.go:246  [-] 1  random seed: -3010804462599978228
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/a38679e165d5c98145544b9bfab9e765/logTestMuxRangeFeedDoesNotStallOnError3232800795
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    dist_sender_rangefeed_test.go:389: condition failed to evaluate within 45s: from dist_sender_rangefeed_test.go:395: at least one non-local range required for test
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a38679e165d5c98145544b9bfab9e765/logTestMuxRangeFeedDoesNotStallOnError3232800795
--- FAIL: TestMuxRangeFeedDoesNotStallOnError (48.14s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMuxRangeFeedDoesNotStallOnError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvclient/kvcoord.TestMuxRangeFeedDoesNotStallOnError [failed](https://tanzanite.cluster.engflow.com/invocations/default/c0dc44e2-336d-49fc-b523-6d7a17138789?testReportRun=11&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZDprdmNvb3JkX3Rlc3Q=) on master @ [bcc993d796d03664604bf695e38fd5644d0bc952](https://github.com/cockroachdb/cockroach/commits/bcc993d796d03664604bf695e38fd5644d0bc952):


```
=== RUN   TestMuxRangeFeedDoesNotStallOnError
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestMuxRangeFeedDoesNotStallOnError2853565186
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    dist_sender_rangefeed_test.go:389: condition failed to evaluate within 45s: from dist_sender_rangefeed_test.go:395: at least one non-local range required for test
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestMuxRangeFeedDoesNotStallOnError2853565186
--- FAIL: TestMuxRangeFeedDoesNotStallOnError (50.19s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=11</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMuxRangeFeedDoesNotStallOnError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvclient/kvcoord.TestMuxRangeFeedDoesNotStallOnError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17982038?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17982038?buildTab=artifacts#/) on master @ [b1474fa887606008960634b571cf4501efb6281b](https://github.com/cockroachdb/cockroach/commits/b1474fa887606008960634b571cf4501efb6281b):


```
=== RUN   TestMuxRangeFeedDoesNotStallOnError
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/a38679e165d5c98145544b9bfab9e765/logTestMuxRangeFeedDoesNotStallOnError2143272062
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    dist_sender_rangefeed_test.go:389: condition failed to evaluate within 45s: from dist_sender_rangefeed_test.go:395: at least one non-local range required for test
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a38679e165d5c98145544b9bfab9e765/logTestMuxRangeFeedDoesNotStallOnError2143272062
--- FAIL: TestMuxRangeFeedDoesNotStallOnError (47.76s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMuxRangeFeedDoesNotStallOnError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvclient/kvcoord/BUILD.bazel
+++ b/pkg/kv/kvclient/kvcoord/BUILD.bazel
@@ -167,6 +167,7 @@ go_test(
         "//pkg/config/zonepb",
         "//pkg/gossip",
         "//pkg/gossip/simulation",
+        "//pkg/jobs",
         "//pkg/keys",
         "//pkg/kv",
         "//pkg/kv/kvbase",

--- a/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go
@@ -15,6 +15,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
@@ -345,7 +346,10 @@ func TestMuxRangeFeedDoesNotStallOnError(t *testing.T) {
 					StreamClientInterceptor: streamInterceptor,
 				},
 			},
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 		},
+		// Scan like a bat out of hell to ensure down-replication happens quickly.
+		ScanInterval: 50 * time.Millisecond,
 	}
 
 	tc := testcluster.StartTestCluster(t, numServers, base.TestClusterArgs{ServerArgs: serverArgs})
@@ -362,6 +366,8 @@ func TestMuxRangeFeedDoesNotStallOnError(t *testing.T) {
 	sqlDB.ExecMultiple(t,
 		`SET CLUSTER SETTING kv.rangefeed.enabled = true`,
 		`SET CLUSTER SETTING kv.closed_timestamp.target_duration='100ms'`,
+		`SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '50ms'`,
+		`SET CLUSTER SETTING kv.rangefeed.closed_timestamp_refresh_interval = '50ms'`,
 		`ALTER DATABASE defaultdb CONFIGURE ZONE USING num_replicas = 1`,
 		`CREATE TABLE foo (key INT PRIMARY KEY)`,
 	)
@@ -373,7 +379,7 @@ func TestMuxRangeFeedDoesNotStallOnError(t *testing.T) {
 		sqlDB.QueryRow(t,
 			"SELECT sum(cardinality(replicas)) FROM [SHOW RANGES FROM TABLE foo WITH DETAILS]").
 			Scan(&replicaCount)
-		if replicaCount < 3 {
+		if replicaCount != 1 {
 			return errors.Newf("too many replicas: %d", replicaCount)
 		}
 		return nil
@@ -382,16 +388,16 @@ func TestMuxRangeFeedDoesNotStallOnError(t *testing.T) {
 	sqlDB.ExecMultiple(t,
 		`INSERT INTO foo (key) SELECT * FROM generate_series(1, 100)`,
 		`ALTER TABLE foo SPLIT AT (SELECT * FROM generate_series(10, 90, 10))`,
-		`ALTER TABLE foo SCATTER`,
 	)
 
 	// We scatter and wait until we have at least one range without replicas on n1.
 	testutils.SucceedsSoon(t, func() error {
+		sqlDB.Exec(t, `ALTER TABLE foo SCATTER`)
 		var nonLocalCount int
 		sqlDB.QueryRow(t,
 			"SELECT count(1) FROM [SHOW RANGES FROM TABLE foo WITH DETAILS] WHERE array_position(replicas, 1) IS NULL").
 			Scan(&nonLocalCount)
-		if nonLocalCount <= 1 {
+		if nonLocalCount == 0 {
 			return errors.New("at least one non-local range required for test")
 		}
 		return nil

